File: leetcode/apple/group_anagrams.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> table;
        string sorted;
        for (auto &s : strs) {
            sorted = s;
            sort(sorted.begin(), sorted.end());
            table[sorted].push_back(s);
        }

        vector<vector<string>> groups;
        for (auto it : table) {
            groups.push_back(it.second);
        }

        return groups;
    }
};

void print(vector<vector<string>> groups) {
    for (vector<string> &group : groups) {
        for (string &s : group) {
            cout << s << ' ';
        }
        cout << endl;
    }
}

int main() {
    Solution S;

    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};

    auto groups = S.groupAnagrams(strs);
    print(groups);

    return 0;
}

File: leetcode/neetcode_roadmap/group_anagrams.py
class Solution:
    def groupAnagrams(self, words: list[str]) -> list[list[str]]:
        sorted_words = [''.join(sorted(list(w))) for w in words]

        groups = {}
        for i, w in enumerate(sorted_words):
            if w in groups:
                groups[w].append(words[i])
            else:
                groups[w] = [words[i]]

        return list(groups.values())


S = Solution()

strs = ["eat","tea","tan","ate","nat","bat"]
print(S.groupAnagrams(strs))
